[fix](type) Data precision is lost when converting DOUBLE type data to DECIMAL (#17191)

herry2038 opened the pull request:
# Proposed changes

Issue Number: close #17191

## Problem summary

1. Decimal template class 's construction method  using strict convernion from double to int ，it lose data precision , using round instead .
2. DecimalV2 's value 's scale is always 9, so when output to mysqlwriter, the to string function will lose data precision too, fix it .
3. Add regression test regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy for the problem.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

Commit:
[fix] Data precision is lost when converting DBOUBLE type data to DECIMAL (#17191)

1. Fix precision lost when convert to double to decimal.
2. Fix the same problem for decimalv3.

Commit:
[fix] Data precision is lost when converting DBOUBLE type data to DECIMAL #17191

1. Fix bug when converting DOUBLE to DECIMAL;
2. Fix bug when converting DOUBLE to DECIMALV3;

Commit:
[fix] 17191
regression test suite update.
1. add convert 10.0245 to decimal(10,3)

Yukang-Lian commented on be/src/runtime/decimalv2_value.cpp:
```diff
@@ -373,7 +373,19 @@ std::string DecimalV2Value::to_string(int scale) const {
         }
     } else {
-        frac_val = frac_val / SCALE_TRIM_ARRAY[scale];
+        // frac_val = frac_val / SCALE_TRIM_ARRAY[scale];
```
please remove useless comment.

herry2038 replied:
Done

Yukang-Lian commented on be/src/vec/core/types.h:
```diff
@@ -308,9 +308,20 @@ struct Decimal {
-    DECLARE_NUMERIC_CTOR(Float64)
+    // FIX 17191
+    //DECLARE_NUMERIC_CTOR(Float32)
+    //DECLARE_NUMERIC_CTOR(Float64)
```
ditto

herry2038 replied:
Thanks, done!

Yukang-Lian commented on regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy:
```diff
@@ -0,0 +1,71 @@
+suite("test_cast_double_to_decimal") {
+    def db = "test_double"
+    sql "CREATE DATABASE IF NOT EXISTS ${db}"
+    sql "use ${db}"
```
We don't need to use these 3 lines. Actually, regression test framework will create db automaticlly. You just need to create table.
```suggestion
     def tableName = "tbl_test_xxx"
    sql """DROP TABLE IF EXISTS ${tableName}"""
    sql """
            CREATE TABLE IF NOT EXISTS ${tableName} (xxx)
            """
           
```

Yukang-Lian commented on regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy:
```diff
@@ -0,0 +1,71 @@
+	sql "insert into test5 values(14,-15.999999995)"
+	sql "insert into test5 values(15,-15.9999)"
+	sql "insert into test5 values(16,-0.9999)"
+	sql "insert into test5 values(17,-15.0245)"
```
maybe you can use one line `insert` statement.

Yukang-Lian commented on regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy:
```diff
@@ -0,0 +1,71 @@
+			//assertEquals(keyAndExcpetedValue[i+1], result[0][i].toString())
+			assertEquals(keyAndExcpetedValue[i+1], result[0][i].toString())
+		}
+	}	
```
we don't need to use `assertequals`, use `qt_select` statement and generate `.out` file.
reference https://doris.incubator.apache.org/zh-CN/community/developer-guide/regression-testing

Commit:
[fix] Data precision is lost when converting DBOUBLE type data to DEC…
…IMAL #17191

After clang-format

Commit:
[fix] Data precision is lost when converting DBOUBLE type data to DECMAL #17191

Repaire UT test failed!
1. Fix DeciamV2Value::to_buffer function.
2. Fix UT test value for
DecimalV2Value::to_string("999999999999999999.999",3)

Commit:
[fix] Data precision is lost when converting DOUBLE type data to DECIMAL (#17191)

1. Remove unneed commension.
2. Optimize regression test and generate outer file.

herry2038 commented:
@Yukang-Lian 帮我再看看，谢谢！

Yukang-Lian reviewed (dismissed):
Good job! LGTM

Yukang-Lian commented:
@Gabriel39 PTAL

Gabriel39 reviewed (dismissed):
LGTM

Gabriel39 commented:
run buildall

Yukang-Lian commented:
run buildall

Gabriel39 reviewed (approved):
LGTM

Gabriel39 commented:
Merged. Thanks for your contribution!
